### ISSUE TITLE
Slow training

### ISSUE THREAD

**Issue by Colanim** [microsoft/DeepSpeed#255]
I'm trying to train BART from HuggingFace library.

But the training is extremely slow, on 4 GPU. Logs show this :

>[INFO 2020-06-08 05:26:19] 0/2650, SamplesPerSec=13.19749106827976
[INFO 2020-06-08 05:26:34] 0/2700, SamplesPerSec=13.198177991591365

It says 13 samples per sec, but if we look at the steps number and log time, we can see it takes 15 sec to do 50 samples =>  **3.33 samples per sec**.

**Why is the training so slow and why the logs are showing contradicted information ?**

**Comment by jeffra** [comment]
Can you enable wall_clock_breakdown in your deepspeed json config? https://www.deepspeed.ai/docs/config-json/#logging

This breakdown should give us some additional information on where the time is being spent.

**Comment by Colanim** [comment]
Thanks for the lightning-fast answer !

Here is some logs with the `wall_clock_breakdown` enabled :

>[INFO 2020-06-08 06:02:19] 0/450, SamplesPerSec=7.458824487454903
time (ms) | forward_microstep: 193.65 | backward_microstep: 329.25 | backward_inner_microstep: 329.14 | backward_allreduce_microstep: 0.05 | step_microstep: 0.92
time (ms) | forward_microstep: 203.21 | backward_microstep: 323.86 | backward_inner_microstep: 323.81 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 195.00 | backward_microstep: 325.61 | backward_inner_microstep: 325.56 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 186.87 | backward_microstep: 330.72 | backward_inner_microstep: 330.64 | backward_allreduce_microstep: 0.04 | step_microstep: 0.05
time (ms) | forward_microstep: 183.16 | backward_microstep: 334.45 | backward_inner_microstep: 334.34 | backward_allreduce_microstep: 0.05 | step_microstep: 0.08
time (ms) | forward_microstep: 186.46 | backward_microstep: 338.38 | backward_inner_microstep: 338.33 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 199.31 | backward_microstep: 335.63 | backward_inner_microstep: 335.56 | backward_allreduce_microstep: 0.03 | step_microstep: 0.05
time (ms) | forward_microstep: 192.47 | backward_microstep: 338.06 | backward_inner_microstep: 337.95 | backward_allreduce_microstep: 0.05 | step_microstep: 0.07
time (ms) | forward_microstep: 192.41 | backward_microstep: 333.76 | backward_inner_microstep: 333.70 | backward_allreduce_microstep: 0.03 | step_microstep: 0.04
time (ms) | forward_microstep: 189.00 | backward_microstep: 338.92 | backward_inner_microstep: 338.81 | backward_allreduce_microstep: 0.05 | step_microstep: 0.07
time (ms) | forward_microstep: 186.59 | backward_microstep: 341.10 | backward_inner_microstep: 341.04 | backward_allreduce_microstep: 0.03 | step_microstep: 0.03
time (ms) | forward_microstep: 193.71 | backward_microstep: 343.33 | backward_inner_microstep: 343.22 | backward_allreduce_microstep: 0.05 | step_microstep: 0.07
time (ms) | forward_microstep: 193.71 | backward_microstep: 341.28 | backward_inner_microstep: 341.21 | backward_allreduce_microstep: 0.03 | step_microstep: 0.04
time (ms) | forward_microstep: 201.06 | backward_microstep: 333.03 | backward_inner_microstep: 332.95 | backward_allreduce_microstep: 0.03 | step_microstep: 0.05
time (ms) | forward_microstep: 196.37 | backward_microstep: 850.88 | backward_inner_microstep: 326.98 | backward_allreduce_microstep: 523.86 | step_microstep: 193.84
time (ms) | forward: 3079.24 | backward: 5859.61 | backward_inner: 5334.38 | backward_allreduce: 524.09 | step: 195.15
time (ms) | forward_microstep: 193.11 | backward_microstep: 319.56 | backward_inner_microstep: 319.51 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 182.26 | backward_microstep: 338.71 | backward_inner_microstep: 338.64 | backward_allreduce_microstep: 0.03 | step_microstep: 0.04
time (ms) | forward_microstep: 190.97 | backward_microstep: 341.18 | backward_inner_microstep: 341.11 | backward_allreduce_microstep: 0.03 | step_microstep: 0.04
time (ms) | forward_microstep: 203.62 | backward_microstep: 332.63 | backward_inner_microstep: 332.58 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 195.21 | backward_microstep: 331.61 | backward_inner_microstep: 331.56 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 204.40 | backward_microstep: 325.38 | backward_inner_microstep: 325.25 | backward_allreduce_microstep: 0.05 | step_microstep: 0.07
time (ms) | forward_microstep: 188.99 | backward_microstep: 343.44 | backward_inner_microstep: 343.36 | backward_allreduce_microstep: 0.03 | step_microstep: 0.04
time (ms) | forward_microstep: 187.01 | backward_microstep: 328.58 | backward_inner_microstep: 328.49 | backward_allreduce_microstep: 0.03 | step_microstep: 0.04
time (ms) | forward_microstep: 189.68 | backward_microstep: 331.94 | backward_inner_microstep: 331.86 | backward_allreduce_microstep: 0.03 | step_microstep: 0.05
time (ms) | forward_microstep: 193.26 | backward_microstep: 332.35 | backward_inner_microstep: 332.30 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 192.74 | backward_microstep: 330.04 | backward_inner_microstep: 329.98 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 191.39 | backward_microstep: 334.16 | backward_inner_microstep: 334.04 | backward_allreduce_microstep: 0.05 | step_microstep: 0.07
time (ms) | forward_microstep: 185.21 | backward_microstep: 331.86 | backward_inner_microstep: 331.81 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 194.69 | backward_microstep: 357.35 | backward_inner_microstep: 357.30 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 199.06 | backward_microstep: 332.70 | backward_inner_microstep: 332.58 | backward_allreduce_microstep: 0.05 | step_microstep: 0.07
time (ms) | forward_microstep: 195.33 | backward_microstep: 585.38 | backward_inner_microstep: 331.97 | backward_allreduce_microstep: 253.37 | step_microstep: 205.53
time (ms) | forward: 3086.26 | backward: 5596.61 | backward_inner: 5342.04 | backward_allreduce: 253.53 | step: 205.91
time (ms) | forward_microstep: 204.31 | backward_microstep: 347.69 | backward_inner_microstep: 347.64 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 203.15 | backward_microstep: 353.87 | backward_inner_microstep: 353.74 | backward_allreduce_microstep: 0.06 | step_microstep: 0.08
time (ms) | forward_microstep: 202.74 | backward_microstep: 338.03 | backward_inner_microstep: 337.90 | backward_allreduce_microstep: 0.06 | step_microstep: 0.08
time (ms) | forward_microstep: 194.12 | backward_microstep: 334.57 | backward_inner_microstep: 334.45 | backward_allreduce_microstep: 0.05 | step_microstep: 0.07
time (ms) | forward_microstep: 199.06 | backward_microstep: 357.45 | backward_inner_microstep: 357.37 | backward_allreduce_microstep: 0.04 | step_microstep: 0.05
time (ms) | forward_microstep: 206.16 | backward_microstep: 353.00 | backward_inner_microstep: 352.91 | backward_allreduce_microstep: 0.04 | step_microstep: 0.06
time (ms) | forward_microstep: 201.97 | backward_microstep: 334.28 | backward_inner_microstep: 334.16 | backward_allreduce_microstep: 0.05 | step_microstep: 0.07
time (ms) | forward_microstep: 189.21 | backward_microstep: 336.03 | backward_inner_microstep: 335.97 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 195.16 | backward_microstep: 338.33 | backward_inner_microstep: 338.28 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 194.98 | backward_microstep: 337.91 | backward_inner_microstep: 337.86 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 194.96 | backward_microstep: 334.13 | backward_inner_microstep: 334.02 | backward_allreduce_microstep: 0.05 | step_microstep: 0.07
time (ms) | forward_microstep: 199.71 | backward_microstep: 325.73 | backward_inner_microstep: 325.64 | backward_allreduce_microstep: 0.05 | step_microstep: 0.05
time (ms) | forward_microstep: 191.91 | backward_microstep: 350.42 | backward_inner_microstep: 350.33 | backward_allreduce_microstep: 0.04 | step_microstep: 0.06
time (ms) | forward_microstep: 193.48 | backward_microstep: 329.82 | backward_inner_microstep: 329.76 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 190.36 | backward_microstep: 330.53 | backward_inner_microstep: 330.45 | backward_allreduce_microstep: 0.03 | step_microstep: 0.05
time (ms) | forward_microstep: 190.08 | backward_microstep: 442.58 | backward_inner_microstep: 334.73 | backward_allreduce_microstep: 107.81 | step_microstep: 192.72
time (ms) | forward: 3150.64 | backward: 5544.11 | backward_inner: 5434.84 | backward_allreduce: 108.06 | step: 193.21
time (ms) | forward_microstep: 200.72 | backward_microstep: 349.60 | backward_inner_microstep: 349.53 | backward_allreduce_microstep: 0.03 | step_microstep: 0.04
time (ms) | forward_microstep: 205.10 | backward_microstep: 351.63 | backward_inner_microstep: 351.58 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
time (ms) | forward_microstep: 199.97 | backward_microstep: 348.74 | backward_inner_microstep: 348.68 | backward_allreduce_microstep: 0.02 | step_microstep: 0.03
[INFO 2020-06-08 06:02:47] 0/500, SamplesPerSec=7.421826531600121

_`SamplesPerSec` is now 7 because I used 2 GPU instead of 4_

---

**I'm not really sure how to interpret this result. Anything seems out-of-place ?**

**Comment by ShadenSmith** [comment]
Hi @Colanim, it looks like you are using gradient accumulation from those logs. You can see that the `allreduce` and `step` are close to zero for more steps - they are amortized over time by accumulating local gradients for multiple steps before reducing them and then updating model steps.

You can see from this line that only about a small amount of time is spent in the communication step (allreduce) since multiple forward/backward passes have been performed:
```
time (ms) | forward: 3079.24 | backward: 5859.61 | backward_inner: 5334.38 | backward_allreduce: 524.09 | step: 195.15
```
To accelerate training, if memory permits you can try to use a larger microbatch size to increase the amount of work you perform in each forward/backward pass.

**Comment by Colanim** [comment]
Thank you for your answer.

Unfortunately, my GPU have a small amount of memory : I can't put more sample in a batch.

I'll close if nothing seems out of place. Thanks again for your answer !
